### PR TITLE
cmake: explain BUILD_VERSION and KERNEL_VERSION_*

### PR DESCRIPTION
**Pull request by marc-hb**
- Add examples for the latter.
- Point at each other and highlight how independent they are from each
  other.
- State their inputs and outputs in plain English.
- Fix "git describe" error message giving the wrong impression that
  everyone cares about BUILD_VERSION. Only the boot banner cares now.

Signed-off-by: Marc Herbert <marc.herbert@intel.com>

### CODE REVIEW

File: cmake/version.cmake [hunk 1]
```diff
@@ -1,5 +1,33 @@
...
+# -------------
+#
+# Inputs:
+#
+#   ``*VERSION*`` and other constants set by
+#   maintainers in ``${ZEPHYR_BASE}/VERSION``
+#
+# Outputs with examples::
+#
+#   PROJECT_VERSION          1.14.99.42
```
Comment (pfalcon) [inline cmake/version.cmake]: If the idea was "to explain everything", then I'd like to point that the list below doesn't explain what is "42" at the end.

File: cmake/git.cmake [hunk 2]
```diff
@@ -1,6 +1,20 @@
...
-# Use git if it is installed, to set BUILD_VERSION
+#.rst:
+# git.cmake
+# ---------
+# If the user didn't already define BUILD_VERSION then try to initialize
+# it with the output of "git describe". Warn but don't error if
+# everything fails and leave BUILD_VERSION undefined.
+#
+# Inputs:
+#   ``BUILD_VERSION`` or the output of ``git describe``
```
Comment (SebastianBoe) [inline cmake/git.cmake]: How is

the output of ``git describe``

an input to git.cmake ?

File: cmake/version.cmake [hunk 3]
```diff
@@ -56,3 +84,18 @@ unset(MAJOR)
...
+
+  KERNEL_VERSION_STRING: ${KERNEL_VERSION_STRING}
+  KERNEL_VERSION_MAJOR: ${KERNEL_VERSION_MAJOR}
+  KERNEL_VERSION_MINOR: ${KERNEL_VERSION_MINOR}
+  KERNEL_PATCHLEVEL: ${KERNEL_PATCHLEVEL}
+
+  KERNEL_VERSION_NUMBER: ${KERNEL_VERSION_NUMBER}
+  ZEPHYR_VERSION_CODE: ${ZEPHYR_VERSION_CODE}
+  KERNELVERSION: ${KERNELVERSION}
+")
```
Comment (SebastianBoe) [inline cmake/version.cmake]: I don't know about this.

Is the intention to be able to debug? If so, isn't printf-debugging more appropriate?

**Review by SebastianBoe — CHANGES_REQUESTED** [review]
Requesting some changes

File: cmake/version.cmake (hunk 3 above)
Comment (marc-hb) [reply cmake/version.cmake]: By "printf-debugging" do you mean: only locally, nothing in git? So this is of course what I did at first and  I found it much more tedious than usual because of:
- the number of variables and amount of typing involved
- the number of local variables and the time required to weed these out

Comment (SebastianBoe) [reply cmake/version.cmake]: This should be dropped, the noise-to-value ratio does not justify it.

Users that want to see the values of variables can either run cmake under --trace-expand or add printf's.

File: cmake/git.cmake (hunk 2 above)
Comment (SebastianBoe) [reply cmake/git.cmake]: The output of git describe is not an input to git.cmake, it is an internal process of git.cmake.

If we had run git describe before git.cmake, then you could say it was an input.

File: cmake/version.cmake (hunk 3 above)
Comment (marc-hb) [reply cmake/version.cmake]: OK will drop this. This code was useful to "bootstrap" the sample output at the top but less so now.

File: cmake/git.cmake (hunk 2 above)
Comment (marc-hb) [reply cmake/git.cmake]: I don't see why who runs git describe and when matters. The terms "input" and "output" are standard for  internal calls too:
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/out-parameter-modifier
https://docs.swift.org/swift-book/LanguageGuide/Functions.html
https://www.learncpp.com/cpp-tutorial/73-passing-arguments-by-reference/
etc.

In fact CMake's documentation for execute_process() uses the terms "input" and "output" too:
https://cmake.org/cmake/help/v3.14/command/execute_process.html?highlight=output

Comment (SebastianBoe) [reply cmake/git.cmake]: I find the documentation causes more confusion than than it clarifies, but perhaps this is subjective.

Perhaps "the output of 'git describe'" will clarify things for some users.

This is OK.

Comment (marc-hb) [reply cmake/git.cmake]: I wrote the "inputs" and "outputs" section in git.cmake for consistency with version.cmake but now I see they've become mostly redundant with the plain English sentence just before. I'll drop them; less is more. I assume the first sentence is OK.
The last sentence with the reference to the (independent!) KERNEL_VERSION_* is the most important IMHO

File: cmake/version.cmake (hunk 1 above)
Comment (marc-hb) [reply cmake/version.cmake]: Despite the "42" example (changed, thx) the goal is to explain but not "everything". This is more about not having to read the whole cmake file or git grep the entire source code to filter out the many temporary variables and find the names that matters. The added comment points at the VERSION file where TWEAK is defined so this is hopefully enough. I considered mentioning TWEAK here but @sebastianboe just helped me trim this down significantly (thx) so I'm reluctant to add stuff again.

**Review by SebastianBoe — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
Nice addition, thanks!
